jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/ssl/SslConnection.java

```diff
@@ -121,13 +121,16 @@ private enum FlushState
     private RetainableByteBuffer _decryptedInput;
     private RetainableByteBuffer _encryptedInput;
     private RetainableByteBuffer _encryptedOutput;
-    private boolean _renegotiationAllowed;
-    private int _renegotiationLimit = -1;
     private boolean _closedOutbound;
-    private boolean _requireCloseMessage;
     private FlushState _flushState = FlushState.IDLE;
     private FillState _fillState = FillState.IDLE;
     private boolean _underflown;
+
+    // Those fields have getters/setters not protected by the lock.
+    private boolean _renegotiationAllowed;
+    private int _renegotiationLimit = -1;
+    private boolean _requireCloseMessage;
```

Comment:
should they be volatile or are they effectively final?   Can you say why it is ok not to lock in the comment?

I think it is because these are effectively final, so perhaps lock in the setters and check if onOpen has been called, and if so then throw ISE?

Reply:
All three are set in `SslClientConnectionFactory.customize()`, which is called right after the `Connection`'s creation and before the `Connection` is made visible to other threads, so memory visibility of these fields is fine without the need to mark them `volatile`.

`_renegotiationAllowed` and `_requireCloseMessage` are set-once fields, but `_renegotiationLimit` is updated as the connection is in use; so the former two are effectively final but not the latter. And since `_renegotiationLimit` is only updated by a single thread at a time, its memory visibility is fine too.

Checking in `onOpen()` isn't necessary, it's an extra safety net that requires an extra volatile boolean. Let me add it as well as some commenting about.